### ISSUE TITLE
module init macro

### ISSUE THREAD

**Issue by Qard** [theduke/quickjs-rs#68]
It'd be cool to have a fancy macro to build the extern function required to create a native module that can be imported by the `qjs` CLI. With `qjs` you can build and import native modules from `*.so` files by simply having a function something like:

```c
JSModuleDef* js_init_module_qjsc_my_module(JSContext* ctx, const char* module_name);
```

Then in JS you just do:

```js
import * from 'my_module.so'
```

**Comment by theduke** [comment]
Not quite sure I follow:

you would like to write Rust code that can be easily compiled to qjs CLI compatible native modules?

**Comment by Qard** [comment]
Yes, which requires an entrypoint function with that signature. It'd be nice to have a macro that lets you write a function with the rust-wrapped types and the macro will wrap it appropriately to expose the C types.

**Comment by theduke** [comment]
That would indeed be convenient.

I guess the interesting questions are around how to provide variables, functions, ... from Rust in a ergonomic way.

I'll probably work on a Webassembly runtime and some other features like async integration first, but contributions are definitely welcome.

It would definitely help if you could sketch out how you would expect code to look like.

**Comment by Qard** [comment]
I might get around to contributing it myself, if I can find the time at some point and no one else gets to it first. My thinking is something like:

```rust
use quick_js::Context;

#[quickjs_init_module(my_module)]
fn init_module(context: Context, name: String) -> Module {
  Module::new(context, name, |module| {
    module.add_callback("add", |a: i32, b: i32| a + b).unwrap();
  })
}
```
For reference, here's what it looks like in C: https://github.com/Qard/quickjs-glfw/blob/master/glfw.c#L178-L188
